base/irrationals.jl

```diff
@@ -13,7 +13,22 @@ promote_rule{s,T<:Number}(::Type{Irrational{s}}, ::Type{T}) = promote_type(Float
 convert(::Type{AbstractFloat}, x::Irrational) = Float64(x)
 convert(::Type{Float16}, x::Irrational) = Float16(Float32(x))
 convert{T<:Real}(::Type{Complex{T}}, x::Irrational) = convert(Complex{T}, convert(T,x))
-convert{T<:Integer}(::Type{Rational{T}}, x::Irrational) = convert(Rational{T}, Float64(x))
+
+@generated function convert{T<:Integer}(::Type{Rational{T}}, x::Irrational)
```

Comment:
Returning a value is okay (it just gets spliced in as an expression). 

The aim of the `@generated` is to get it to perform the calculation at compile time, as the result will always be constant. This is exactly the same as the usage in the function directly below.

Technically, an `@pure` annotation would be sufficient , once #14324 is sorted out.